Studies/source/GenStudy.cxx

```diff
@@ -84,148 +97,124 @@ class GenStudy {
-                throw exception("passed TEfficiency objects do not have consistent bin contents");
-            TEfficiency eff(passed, total);
-
-            eff.SetConfidenceLevel(oneSigma);
-            eff.SetStatisticOption(TEfficiency::kFCP);
-
-            if(print_info)
-                std::cout << info_name <<"=" << eff.GetEfficiency(2) << " error_up=" << eff.GetEfficiencyErrorUp(2)
-                          << " error_low=" << eff.GetEfficiencyErrorLow(2) << std::endl;
-            else{
-
-                std::ostringstream ss_name;
-                ss_name  << channel << " " << hist_name;
-                std::string name = ss_name.str();
-
-                eff.SetTitle(name.c_str());
-                canvas.Clear();
-                const std::vector<int> x_bins = {25,30,35,40,45,50,60,70,80,90,100,120,160,200,300,400,600,800,1000};
-                double min = 1, max = 0;
-                for(int n = 1; n <= passed.GetNbinsX(); ++n) {
-                    min = std::min(eff.GetEfficiency(n), min);
-                    max = std::max(eff.GetEfficiency(n), max);
-                    empty_hist.GetXaxis()->SetBinLabel(n, ToString(x_bins.at(n)).c_str());
-                    empty_hist.GetXaxis()->SetTitle(("Pt vis [GeV]"));
-                    empty_hist.GetYaxis()->SetTitle(("Efficiency"));
-                    empty_hist.GetYaxis()->SetTitleOffset(1.5f);
-                    empty_hist.SetLabelSize(0.04f);
-                }
-
-                min = min > 0.2 ? min / range_sf : 0;
-                empty_hist.GetYaxis()->SetRangeUser(min, max * range_sf);
-                empty_hist.Draw();
-                eff.Draw("SAME P");
-                empty_hist.Write();
-                canvas.SetLogx();
-
-                empty_hist.SetTitle(name.c_str());
-                eff.SetDirectory(pFile);
-                empty_hist.SetDirectory(pFile);
-                eff.Write();
-                pFile->Write();
-
-                canvas.Print((args.outputFile()+".pdf").c_str(), ("Title:"+name).c_str());
-                canvas.Clear();
-            }
+        constexpr static double oneSigma = 0.682689492137;
+
+        if(!TEfficiency::CheckConsistency(passed, total))
+            throw exception("passed TEfficiency objects do not have consistent bin contents.");
+        TEfficiency eff(passed, total);
+
+        eff.SetConfidenceLevel(oneSigma);
+        eff.SetStatisticOption(TEfficiency::kFCP);
+
+        if(print_info)
+            std::cout << info_name <<"=" << eff.GetEfficiency(2) << " error_up=" << eff.GetEfficiencyErrorUp(2)
+                      << " error_low=" << eff.GetEfficiencyErrorLow(2)  << std::endl;
+        else{
+            std::ostringstream ss_name;
+            ss_name  << channel << "_" << hist_name;
+            std::string name = ss_name.str();
+
+            eff.SetTitle(name.c_str());
+            eff.SetDirectory(output.get());
+            eff.Write();
```

Comment:
SetDirectory and Write are redundant, since WriteObject is called below

Reply:
I don't understand why it creates a copy of the eff plots on the new_tuple rather than only output. So that's why I put the SetDirectory to try to avoid that, but it didn't work

Reply:
You need to or change the order in which files are created in the constructor or add histo_file.cd() at the end of the constructor. It should help